x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/DefaultRoleReferenceResolver.java

```diff
@@ -0,0 +1,252 @@
+import org.elasticsearch.xpack.core.security.authz.store.ReservedRolesStore;
+import org.elasticsearch.xpack.core.security.authz.store.RoleReference;
+import org.elasticsearch.xpack.core.security.authz.store.RoleReferenceResolver;
+import org.elasticsearch.xpack.core.security.authz.store.RoleRetrievalResult;
+import org.elasticsearch.xpack.core.security.authz.store.RolesRetrievalResult;
+import org.elasticsearch.xpack.core.security.support.MetadataUtils;
+import org.elasticsearch.xpack.security.authc.ApiKeyService;
+import org.elasticsearch.xpack.security.authc.service.ServiceAccountService;
+
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
+
+import static java.util.function.Predicate.not;
+import static org.elasticsearch.xpack.core.security.SecurityField.DOCUMENT_LEVEL_SECURITY_FEATURE;
+
+public class DefaultRoleReferenceResolver implements RoleReferenceResolver {
+
+    private static final Logger logger = LogManager.getLogger(DefaultRoleReferenceResolver.class);
+    private final DeprecationLogger deprecationLogger = DeprecationLogger.getLogger(DefaultRoleReferenceResolver.class);
+
+    private final RoleProviders roleProviders;
+    private final ApiKeyService apiKeyService;
+    private final ServiceAccountService serviceAccountService;
+    private final XPackLicenseState licenseState;
+    private final ThreadContext threadContext;
+    private final Consumer<Collection<RoleDescriptor>> effectiveRoleDescriptorsConsumer;
+    private final Cache<String, Boolean> negativeLookupCache;
+
+    public DefaultRoleReferenceResolver(
+        RoleProviders roleProviders,
+        ApiKeyService apiKeyService,
+        ServiceAccountService serviceAccountService,
+        Cache<String, Boolean> negativeLookupCache,
+        XPackLicenseState licenseState,
+        ThreadContext threadContext,
+        Consumer<Collection<RoleDescriptor>> effectiveRoleDescriptorsConsumer
+    ) {
+        this.roleProviders = roleProviders;
+        this.apiKeyService = Objects.requireNonNull(apiKeyService);
+        this.serviceAccountService = Objects.requireNonNull(serviceAccountService);
+        this.licenseState = Objects.requireNonNull(licenseState);
+        this.threadContext = threadContext;
+        this.effectiveRoleDescriptorsConsumer = Objects.requireNonNull(effectiveRoleDescriptorsConsumer);
+        this.negativeLookupCache = negativeLookupCache;
+    }
+
+    @Override
+    public void resolveNamedRoleReference(
+        RoleReference.NamedRoleReference namedRoleReference,
+        ActionListener<RolesRetrievalResult> listener
+    ) {
+        final Set<String> roleNames = Set.copyOf(new HashSet<>(List.of(namedRoleReference.getRoleNames())));
```

Comment:
Why the nested collections?
```suggestion
        final Set<String> roleNames = Set.of(namedRoleReference.getRoleNames());
```

Reply:
Unfortunately, `namedRoleReference.getRoleNames()` can have duplicate names and `Set.of` does not like it. But the outermost `Set.copyOf` is not necessary (I just added it out of habit) and I can drop it if you prefer.